blogs/2018/03/text-buffer-reimplementation.md

```diff
@@ -0,0 +1,277 @@
+So I've set out to benchmark this method in various scenarios:
+
+ * call `getLineContent` for all lines after doing 1000 random edits
+ * call `getLineContent` for all lines after doing 1000 sequential inserts
+ * read 10 distinct line windows after doing 1000 random edits
+ * read 10 distinct line windows after doing 1000 sequential inserts.
+
+<center>
+<img src="./read.png" style="width: 800px" alt="Read all lines after random edits">
+</center>
+
+TADA, we found the achilles heel of Piece Tree. A large file, with 1000s of edits, will lead to thousands or tens of thousands of nodes. Even though looking up a line is `O(log N)`, where `N` is the number of nodes, that is significantly more than `O(1)` which the line array enjoyed.
+
+However, we are talking about microseconds for each `getLineContent` call, and it is not something we are concerned with at this time. Most of `getLineContent` calls are from View Rendering and Tokenizer, and the post processes of line contents are much more time consuming. DOM construction and rendering or tokenization of a view port usually takes tens of milliseconds, in which `getLineContent` only accounts for less than 1%. But we are considering eventually implementing a *normalization* step, where we would recreate buffers and nodes if certain conditions are met (like a very high number of nodes).
+
+## Conclusion and Gotchas
+
+Piece Tree outperforms Line Array in most scenarios, with the exception of line based lookup, which was to be expected.
+
+### Lessons learned
+
+* The most important lesson this reimplementation taught me is to **always do real world profiling**, every time I did that I found that my assumptions about which methods would be hot did not match the reality, For example, I focused a lot on tuning three atomic operations, `insert`, `delete` and `search` when I started the Piece Tree implementation, but when I integrated it in VSCode, none of those optimizations mattered, as the hottest method was `getLineContent`.
+* **Dealing with `CRLF` or mixed line breaks sequences is a programmer's nightmare**. For every modification, we need to check if the edit splits a CRLF sequence, or if it creates a new CRLF sequence by deleting some content. Dealing with all the possible cases, in the context of a tree, took a lot of attempts until I've reached a solution which is both correct and fast.
+* **GC can easily eat your CPU time**. Our text model and other components used to have the assumption that the buffer is stored in an array so we heavily use `getLineContent` even though sometimes it's not necessary. Say we just want to know the char code of the first character of a line, we used to do a `getLineContent` first and then run `charCodeAt` but with Piece Tree, `getLineContent` creates a substring and after checking the char code, the line substring is thrown away immediately. We are working on adopting better suited methods.
+
+### Why not Native
```

Comment:
Should we say *Why not Native for now*? as JS version is the best for current architecture.